src/sidebar/components/test/tag-editor-test.js

```diff
@@ -1,63 +1,257 @@
+      assert.isTrue(fakeOnEditTags.calledWith({ tags: tagList }));
+      // clears out the suggestions
+      assert.equal(wrapper.find('datalist option').length, 0);
+      // assert the input value is cleared out
+      assert.equal(wrapper.find('input').instance().value, '');
+      // note: focus not tested
+    };
+    /**
+     * Helper function to assert that a tag was correctly not added
+     */
+    const addTagsFail = () => {
+      assert.isTrue(fakeTagsService.store.notCalled);
+      assert.isTrue(fakeOnEditTags.notCalled);
+    };
+
+    it('adds a tag from the input field', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag3';
+
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      addTagsSuccess(wrapper, ['tag1', 'tag2', 'tag3']);
+    });
+
+    it('populate the datalist, then adds a tag from the input field', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag3';
+
+      // simulate `keyup` to populate suggestions list
+      wrapper.find('input').simulate('keyup', { key: 'any' });
+      assert.equal(wrapper.find('datalist option').length, 2);
+
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      addTagsSuccess(wrapper, ['tag1', 'tag2', 'tag3']);
+    });
+
+    it('clears out the <datalist> element after adding a tag', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag3';
+
+      // simulate `keyup` to populate suggestions list
+      wrapper.find('input').simulate('keyup', { key: 'any' });
+      assert.equal(wrapper.find('datalist option').length, 2);
+
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      assert.isFalse(wrapper.find('datalist').exists());
+    });
+
+    it('should not add a tag if the input is empty', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = '';
+
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      addTagsFail();
+    });
+
+    it('should not add a tag if the input is blank space', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = '  ';
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      addTagsFail();
+    });
+
+    it('should not add a tag if its a duplicate of one already in the list', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag1';
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      addTagsFail();
+    });
 
-    beforeEach(function() {
-      onEditTags = sinon.stub();
-      element = util.createDirective(document, 'tag-editor', {
-        onEditTags: { args: ['tags'], callback: onEditTags },
-        tags: ['foo'],
-      });
-      element.ctrl.onTagsChanged();
+    it('adds a tag via keypress `Enter`', () => {
```

Comment:
Low priority, but I feel like these three tests could be rolled into one with test-case objects and a `forEach`. But they're OK as-is.

Reply:
fixed